### ISSUE TITLE
Support certificate verification for HTTPS 

### ISSUE THREAD

**Issue by earlchew** [LMS-Community/slimserver#848]
> As an operator of my Squeezebox server, I want to enable secure HTTPS connections when I opt out of `insecureHTTPS`, so that I have confidence in the authenticity of the remote hosts providing updates and media, and my log files do not fill with repeated warnings.

This issue was discussed in #208 which resulted in 8a129a0cd6b41e9eb29d17f6523712c45f6e8d62 to explicitly set `VERIFY_NONE` to work around an incompatibility with prior use of `0x00`.

A subsequent  commit b787c60c78ab1ef0dfb13b28884a42d24fda6df3 introduced the preference setting `insecureHTTPS` to control with `VERIFY_NONE` would be set explicitly.

Selecting `insecureHTTPS=1` configures `VERIFY_NONE`. Using `insecureHTTPS=0` gives the impression that secure connections will be used, but OpenSSL and IO::Socket::SSL require `VERIFY_PEER` to achieve this. Without explicit configuration:
* https://www.openssl.org/docs/man3.0/man3/SSL_CTX_new.html - _On session estabilishment, by default, no peer credentials verification is done. This must be explicitly requested, typically using [SSL_CTX_set_verify(3)](https://www.openssl.org/docs/man3.0/man3/SSL_CTX_set_verify.html)._
* https://github.com/noxxi/p5-io-socket-ssl/commit/e3888257eda1ad9ba69a2334c2a415576876c6b8 - _The use of the current default SSL_VERIFY_NONE will cause a big warning for clients, unless SSL_verify_mode was explicitly set inside the  application to this insecure value. In the near future the default will be SSL_VERIFY_PEER, and thus causing verification failures in unchanged applications._

The following is proposed to support certificate validation:
* Preserve the present behaviour of setting `VERIFY_NONE` if `insecureHTTPS=1`
* Preserve the present behaviour of relying on defaults if `insecureHTTPS=0`, but also require that neither `SSL_CERT_DIR` and `SSL_CERT_FILE` variables are present in the environment.
* Support configuring `VERIFY_PEER` if `insecureHTTPS=0`, and either one of the`SSL_CERT_DIR` or `SSL_CERT_FILE` environment variables are present (see https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_set_default_verify_paths.html).

**Comment by michaelherger** [comment]
I'm not sure I follow... what kind of message would you get in the log files? What version of IO::Socket::SSL are you using?

**Comment by earlchew** [comment]
> What version of IO::Socket::SSL are you using?
```
commit 7f2b8ec29d43bb62c76ff9630fef86ea8d544231 (HEAD -> tag: 2.060)
Author: Steffen Ullrich <Steffen_Ullrich@genua.de>
Date:   Sun Sep 16 21:14:32 2018 +0200

    release as 2.060
```

> what kind of message would you get in the log files?
```
Slim::Utils::Misc::msg (1266) Warning: [19:42:45.1359] *************************************************
******************
Using the default of SSL_verify_mode of SSL_VERIFY_NONE for client
is deprecated! Please set SSL_verify_mode to SSL_VERIFY_PEER
together with SSL_ca_file|SSL_ca_path for verification.
If you really don't want to verify the certificate and keep the
connection open to Man-In-The-Middle attacks please set
 SSL_verify_mode explicitly to SSL_VERIFY_NONE in your application.
*******************************************************************
```

**Comment by michaelherger** [comment]
Hmm... I happen to be using the same version, but I'm not seeing that message. You are on LMS 8.3 or later, aren't you? 

According to IO::Socket::SSL documentation (https://metacpan.org/pod/IO::Socket::SSL) the default value if non set already is `VERIFY_PEER`. What version of `Net::SSLeay` would you be using? I'm on 1.85.

**Comment by earlchew** [comment]
> ... I happen to be using the same version

Taking a closer look, I see that the log message was removed at https://github.com/openssl/openssl/commit/4936ba4488cf5e89068111253c27038bdcd17d3d, which leaves me a bit confused as to the contents of my log.

```
    1.968 - better support for usable CA path by default
    - new function default_ca which emulates openssl search for default CA path.
      Falls back to Mozilla::CA if no usable CA store is found
    - enforce use of Mozilla::CA on platforms without usable CA store (windows)
    - remove long depreceated support for certs/server-{cert,key}.pem, ca/ and
      certs/my-ca.pem defaults.
```
I suspect the messages were introduced by my earlier use of IO::Socket::SSL 1.94.

I'll take a another look at my configuration and update the ticket.

**Comment by earlchew** [comment]
I have confirmed that I am getting the desired behaviour with IO::Socket::SSL 2.060. This change was introduced at IO::Socket::SSL 1.968 (https://github.com/noxxi/p5-io-socket-ssl/commit/4936ba4488cf5e89068111253c27038bdcd17d3d).

I had become confused as I transitioned from 1.94 to 2.060, conflating the warning messages in the log with the errors establishing the TLS session. Only 1.94 emits the warning message, while 2.060 pays attention to `SSL_CERT_DIR` and `SSL_CERT_FILE`

Also note that IO::Socket::SSL does not yet support `SSL_CERT_PATH` which was introduced after OpenSSL 3.1 (https://github.com/openssl/openssl/commit/606e0426a148034c8c131de9f31f7d3e38be99ea).
